### PR TITLE
Mapping file generation super-powers

### PR DESCRIPTION
**Pull request by prescod**
1. Allow users to inject load declarations into generated mapping file.
2. Allow users to order mapping file declarations
3. Auto-inject "after:" declarations where appropriate.

(the description below is the vision...not every detail is implemented yet)

Still needs CCI side, documentation and tests.

### Snowfakery and Mapping Files

CumulusCI's bulk loader depends on a mapping file to determine the load order of data.

Snowfakery generates this mapping file automatically in the `generate_and_load_from_yaml` job. Occasionally you may need to inject some additional information into the load. For example, you might want to change which API or processing mode to use (REST API, BULK API, Bulk Serial, Bulk Parallel). Or you might want to enforce a certain load order.

You can now do that by specifying a "recipe.load.yml" file to go with your recipe file. The --load-yml command line argument allows that and there is also a file naming convention which does it automatically (TBD).

Here are the declarations you may put into the "recipe.load.yml".

- sf_object: Contact 
  bulk_mode: serial  # same as specifying `bulk_mode` in mapping file

- sf_object: Contact
  load_after: Opportunity  # load contacts after opportunities for some reasoon

- sf_object: Account  # same as specifying `api` in mapping file
  api: rest

### CODE REVIEW

File: snowfakery/cci_mapping_files/declaration_parser.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        if existing_decl:
+            existing_decl.value.append(new_decl.value)
+            return existing_decl
+        else:
+            d = new_decl
+            # start to build a list-based declaration
+            return AtomicDecl(d.sf_object, d.key, [d.value], d.priority, d.merge_rule)
+
+
+class AtomicDecl(T.NamedTuple):
```
Comment (Br4nd0R) [inline snowfakery/cci_mapping_files/declaration_parser.py:29]: You can move this above `MergeRules` and add a doc string :)

File: snowfakery/cci_mapping_files/declaration_parser.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    "anchor_date": MergeRules.use_highest_priority,
+    "load_after": MergeRules.append,
+}
+
+
+class SObjectRuleDeclarationFile(BaseModel):
+    __root__: T.List[SObjectRuleDeclaration]
+
+    @classmethod
+    def parse_from_yaml(cls, data: T.List):
```
Comment (Br4nd0R) [inline snowfakery/cci_mapping_files/declaration_parser.py:81]: change name/semantics of this function

File: snowfakery/cli.py [hunk 3]
```diff
@@ -131,6 +135,11 @@ def int_string_tuple(ctx, param, value=None):
     type=click.File("r"),
     help="Continue generating a dataset where 'continuation-file' left off",
 )
+@click.option(
```
Comment (Br4nd0R) [inline snowfakery/cli.py:138]: Load more than one declaration file?

**Comment by prescod** [comment]
> I think the feature makes sense, but why a separate file instead of including these hints somewhere in the main recipe that the mapping generator knows to look for? Do you anticipate a need to use different sets of hints with the same recipe?

Yes, I do anticipate that e.g. the standard NPSP recipe might work on your org but automation you've installed might make the load order unique to your org.

Or conversely (and perhaps more commonly!) the standard NPSP load file might work but you want to add a custom field to the recipe and therefore create your own.

In the not too distant future I could also imagine you sharing the files between Snowfakery data loads and CCI SQL loads instead of using the succinct Snowfakery format in one place and the verbose and unwieldy mapping file format in the other.

And then in the distant future, perhaps there is some centralized management of them analogous to TypeShed, DefinitelyTyped, SchemaStore.org etc. (or maybe that won't be useful enough)

File: snowfakery/cli.py (hunk 3 above)
Comment (prescod) [reply snowfakery/cli.py:142]: 12e7b03

File: snowfakery/cci_mapping_files/declaration_parser.py (hunk 2 above)
Comment (prescod) [reply snowfakery/cci_mapping_files/declaration_parser.py]: Changed inputs.

File: snowfakery/cci_mapping_files/declaration_parser.py (hunk 1 above)
Comment (prescod) [reply snowfakery/cci_mapping_files/declaration_parser.py]: Done and done.

File: snowfakery/cli.py [hunk 4]
```diff
@@ -344,6 +352,19 @@ def stopping_criteria_from_target_number(target_number):
     return None
 
 
+from pysnooper import snoop
+
+
+@snoop()
```
Comment (Br4nd0R) [inline snowfakery/cli.py]: TIL: `PySnooper`!

**Review by Br4nd0R — APPROVED** [review]
I took a look at the changes since we reviewed this together. All looks good. I really appreciated the addition of the docstrings!

**Comment by davisagli** [comment]
Still needs docs, right?

**Comment by prescod** [comment]
I'm planning to put the docs in the CCI repo.

**Pull request merged**